[CORE-2698] Transaction log store

cb1kenobi opened the pull request:
Adding support for the transaction log store.

Fixes https://github.com/HarperFast/rocksdb-js/issues/217.

kriszyp commented on src/binding/transaction_log.cpp:
```diff
@@ -0,0 +1,190 @@
+
+	// // create a reference to pin the buffer in memory (prevents GC)
+	napi_ref bufferRef = nullptr;
+	NAPI_STATUS_THROWS(::napi_create_reference(env, argv[0], 1, &bufferRef));
```
I think we should copy the contents of the buffer into `data`, keep this as managed memory in C (delete when done) rather than maintaining a JS ref:
* Remove the complexity of managing, tracking, and deleting the JS reference
* Allow the JS buffer to be immediately GC'ed. For very large txns, helps protect against JS OOM, better scalability.
* Or allow for the safe reuse of shared buffer, that can be written over and over. Right now, I think this expects that the buffer won't be modified until the commit takes place, but this shouldn't be assumed, copying the data keeps it safe.

kriszyp commented on src/binding/transaction_log_file_posix.cpp:
```diff
@@ -0,0 +1,97 @@
+		return;
+	}
+
+	if (::fsync(this->fd) < 0) {
```
I think we probably want to use `fdatasync`, a little faster and we don't need to update file metadata with every flush.

cb1kenobi replied:
Ah, yes `fdatasync()`. I'm fine with this, but we need to be comfortable with the tradeoffs. Since `fdatasync()` does not update the log file's mtime, the mtime will be stale until the kernel writeback thread runs and updates the file's mtime. Apparently this can take 5-30 seconds. Should the actualy OS panic or get powered off during this time, then we would have a stale mtime and the rocksdb-js code could potentially delete a transaction log outside the user-defined retention window. We default to 3 days, so odds of this being a problem are really slim. Just something to be aware of.

cb1kenobi replied:
And apparently macOS also doesn't support `fdatasync()`.

kriszyp commented on docs/transaction-log-v1.md:
````diff
@@ -0,0 +1,313 @@
+
+```
++------------------+
+| WOOF Token       | 4 bytes
````
Confirmed! https://github.com/HarperFast/rocksdb-js/pull/238/files#diff-22a3c4e9b693561b0d3d6534ae81880ea0dec05f14fc12b7bbc837a5748e59e5R118

cb1kenobi commented on src/binding/transaction_log_store.h:
```diff
@@ -0,0 +1,143 @@
+	/**
+	 * Writes a batch of transaction log entries to the store.
+	 */
+	void commit(TransactionLogEntryBatch& batch);
```
I love it! I'll make it so.

kriszyp replied:
Excellent, sounds good to me. A transaction can be "written" and then "commited" :+1: .

cb1kenobi commented on docs/transaction-log-v1.md:
```diff
@@ -0,0 +1,313 @@
+
+| Offset | Size | Type    | Field           | Description                                  |
+|--------|------|---------|-----------------|----------------------------------------------|
+| 0      | 8    | uint64  | startTimestamp  | Block timestamp (milliseconds since epoch)   |
```
Done!

kriszyp commented on src/binding/util.cpp:
```diff
@@ -400,4 +415,67 @@ const char* getNapiBufferFromArg(
+		if (now <= last) {
+			now = last + 1;
+		}
+	} while (!lastTimestamp.compare_exchange_strong(last, now, std::memory_order_acq_rel));
```
TIL that `compare_exchange_strong` updates the `last` (expected) value if it doesn't match, completing the update for the loop. Brilliant!

kriszyp commented on src/binding/util.cpp:
```diff
@@ -400,4 +415,67 @@ const char* getNapiBufferFromArg(
+		}
+	} while (!lastTimestamp.compare_exchange_strong(last, now, std::memory_order_acq_rel));
+
+	return static_cast<double>(now) / 1000000000.0;
```
I think you have 3 extra zeros in here, nanoseconds to milliseconds should be dividing by 1000000 (million), right?

cb1kenobi replied:
Fixed!

kriszyp commented on src/binding/transaction_log_store.cpp:
```diff
@@ -0,0 +1,309 @@
+		TransactionLogFile* logFile = nullptr;
+
+		// get the current log file and rotate if needed
+		while (logFile == nullptr && this->currentSequenceNumber) {
```
As far as I can tell, this is not implementing https://github.com/HarperFast/rocksdb-js/pull/238/commits/6dd20de7af5f83c82ec465c5d4e375ded6691cac#diff-2872eef24b542a107a9c6f0b14427d656cf6cd3bef907f9a991b1482d71855acR18:

> A new transaction log file may be started even if the previous file has not reached its maximum size. A new transaction log file should be started if the timestamp is older than 1/8 of the transaction retention period.

(1/4 might be better though)

kriszyp commented on src/binding/util.cpp:
```diff
@@ -400,4 +415,73 @@ const char* getNapiBufferFromArg(
+
+	double last = lastTimestamp.load(std::memory_order_acquire);
+	if (result <= last) {
+		result = std::nextafter(last, std::numeric_limits<double>::infinity());
```
Seems to fail to compile. Maybe just
`result = (double) ((uint64_t)result + 1);` ?

cb1kenobi commented on src/binding/transaction_log_file_posix.cpp:
```diff
@@ -0,0 +1,114 @@
+	}
+
+	// open file for both reading and writing
+	this->fd = ::open(this->path.c_str(), O_RDWR | O_CREAT, 0644);
```
Fixed!

kriszyp commented on src/binding/util.h:
```diff
@@ -373,6 +488,82 @@ struct AsyncWorkHandle {
+	// Read uint64_t in big endian and interpret as double
+	uint64_t bits = readUint64BE(buffer);
+	double value;
+	std::memcpy(&value, &bits, sizeof(double));
```
Can we can cast without a memcpy? `*((double*) &bits)` works, not sure if there is prettier way.

kriszyp commented on src/binding/util.h:
```diff
@@ -373,6 +488,82 @@ struct AsyncWorkHandle {
+}
+
+inline uint64_t readUint64BE(const char* buffer) {
+	return (static_cast<uint64_t>(static_cast<uint8_t>(buffer[0])) << 56) |
```
At some point, maybe we could use `htonl` or `bswap_64` for a single machine operation conversion.

cb1kenobi commented:
@kriszyp I think this is looking good and stable. I believe I have addressed all feedback. I have code reviewed it an tests are passing, so I'm eager to fix any issues you find. Thanks!

kriszyp commented on src/binding/binding.cpp:
```diff
@@ -30,23 +32,23 @@ NAPI_MODULE_INIT() {
 	napi_create_string_utf8(env, rocksdb::GetRocksVersionAsString().c_str(), NAPI_AUTO_LENGTH, &version);
 	napi_set_named_property(env, exports, "version", version);
 
-	[[maybe_unused]] int refCount = ++moduleRefCount;
```
I am curious about this change, as I have been using the same shorthand incrementation/decrementation pattern you are deleting. I thought https://en.cppreference.com/w/cpp/atomic/atomic/operator_arith.html explicitly guarantees `++moduleRefCount` and `moduleRefCount.fetch_add(1) + 1` are identical expressions (and same is true that `--moduleRefCount` and `moduleRefCount.fetch_sub(1) - 1` are guaranteed to be identical) because  `std::atomic` does operator overloading of `++` and `--` to ensure they are atomic. Do you know how they differ or sure they differ?

cb1kenobi commented on src/binding/transaction_log.cpp:
```diff
@@ -0,0 +1,198 @@
+	}
+
+	try {
+		std::unique_ptr<char[]> copyData(new char[size]);
```
Done!

cb1kenobi commented on docs/transaction-log.md:
```diff
@@ -0,0 +1,256 @@
+| Offset | Size | Type    | Field      | Description            |
+|--------|------|---------|------------|------------------------|
+| 0      | 4    | uint32  | token      | Transaction log token  |
+| 4      | 1    | uint8   | version    | Format version         |
```
Done!
